test/integration/otel_test.go

```diff
@@ -0,0 +1,295 @@
+		SpanID  string
+	}
+}
+
+type TraceData struct {
+	TraceID   string
+	Spans     []SpanData
+	Processes map[string]struct {
+		ServiceName string
+	}
+	Warnings []string
+}
+
+// TraceResponse is what we get from the traces API.
+// We always search for a single trace by ID, so this should be length 1.
+type TraceResponse struct {
+	Data []TraceData
+}
+
+func getTraceFromJaeger(t *testing.T, traceID trace.TraceID) TraceData {
+	traceURL := "http://bjaeger:16686/api/traces/" + traceID.String()
+	resp, err := http.Get(traceURL)
+	test.AssertNotError(t, err, "failed to trace from jaeger: "+traceID.String())
+	if resp.StatusCode == http.StatusNotFound {
+		t.Fatalf("jaeger returned 404 for trace %s", traceID)
+	}
+	test.AssertEquals(t, resp.StatusCode, http.StatusOK)
+
+	body, err := io.ReadAll(resp.Body)
+	test.AssertNotError(t, err, "failed to read trace body")
+
+	var parsed TraceResponse
+	err = json.Unmarshal(body, &parsed)
+	test.AssertNotError(t, err, "failed to decode traces body")
+
+	if len(parsed.Data) != 1 {
+		t.Fatalf("expected to get exactly one trace for %s: %v", traceID, parsed)
+	}
+
+	return parsed.Data[0]
+}
+
+type expectedSpans struct {
+	Operation string
+	Service   string
+	Children  []expectedSpans
+}
+
+// isParent returns true if the given span has a parent of ParentID
+// The empty string means no ParentID
+func isParent(parentID string, span SpanData) bool {
+	if len(span.References) == 0 {
+		return parentID == ""
+	}
+	for _, ref := range span.References {
+		// In OpenTelemetry, CHILD_OF is the only reference, but Jaeger supports other systems.
+		if ref.RefType == "CHILD_OF" {
+			return ref.SpanID == parentID
+		}
+	}
+	return false
+}
+
+func missingChildren(traceData TraceData, spanID string, children []expectedSpans) bool {
+	for _, child := range children {
+		if !findSpans(traceData, spanID, child) {
+			// Missing Child
+			return true
+		}
+	}
+	return false
+}
+
+// findSpans checks if the expectedSpan and its expected children are found in traceData
+func findSpans(traceData TraceData, parentSpan string, expectedSpan expectedSpans) bool {
+	for _, span := range traceData.Spans {
+		if !isParent(parentSpan, span) {
+			continue
+		}
+		if expectedSpan.Service != traceData.Processes[span.ProcessID].ServiceName {
+			continue
+		}
+		if expectedSpan.Operation != span.OperationName {
+			continue
+		}
+		if missingChildren(traceData, span.SpanID, expectedSpan.Children) {
+			continue
+		}
+
+		// This span has the correct parent, service, operation, and children
+		return true
+	}
+	fmt.Printf("did not find span %s::%s with parent '%s'\n", expectedSpan.Service, expectedSpan.Operation, parentSpan)
+	return false
+}
+
+// ContextInjectingRoundTripper takes a context that is added to every request sent through this RoundTripper
+// It adds the OpenTelemetry propagation headers.  This is useful for HTTP clients which don't pass through a context,
+// notably including the eggsampler ACME client used in this test.
```

Comment:
It would be useful to note _why_ we need this; namely (as far as I understand it) that we're taking advantage of client-to-server trace propagation in order to be able to find the generated spans by traceID.

Reply:
I've clarified some of the comments here with more information, so hopefully they're now clearer.